ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,313 @@
+// It does not switch to the new profile.
+func (pm *ProfileManager) NewProfile(profile string) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	if slices.Contains(pm.knownProfiles, profile) {
+		return fmt.Errorf("profile %q already exists", profile)
+	}
+	pm.knownProfiles = append(pm.knownProfiles, profile)
+	if err := pm.writeKnownProfiles(); err != nil {
+		return err
+	}
+	// In case one already exists in store, wipe it.
+	return pm.writePrefsToStore(ipn.StateKey("profile-"+profile), emptyPrefs)
+}
+
+// emptyPrefs is the default prefs for a new profile.
+var emptyPrefs = func() ipn.PrefsView {
+	prefs := ipn.NewPrefs()
+	prefs.WantRunning = false
+	return prefs.View()
+}()
+
+// Store returns the StateStore used by the ProfileManager.
+func (pm *ProfileManager) Store() ipn.StateStore {
+	return pm.store
+}
+
+// CurrentPrefs returns a read-only view of the current prefs.
+// It is always valid.
+func (pm *ProfileManager) CurrentPrefs() ipn.PrefsView {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	return pm.prefs
+}
+
+// NewProfileManager creates a new ProfileManager using the provided StateStore.
+// It also loads the list of known profiles from the StateStore.
+// If a state key is provided, it will be used to load the current profile.
+func NewProfileManager(store ipn.StateStore, logf logger.Logf, stateKey ipn.StateKey) (*ProfileManager, error) {
+	return newProfileManagerWithGOOS(store, logf, stateKey, runtime.GOOS)
+}
+
+func newProfileManagerWithGOOS(store ipn.StateStore, logf logger.Logf, stateKey ipn.StateKey, goos string) (*ProfileManager, error) {
+	if stateKey == "" {
+		startKey := ipn.CurrentProfileStateKey
+		stateKeyPrefix := "profile-"
+		if goos == "windows" {
+			// When tailscaled runs on Windows it is not typically run unattended.
+			// So we can't use the profile mechanism to load the profile at startup.
+			startKey = ipn.ServerModeStartKey
+			stateKeyPrefix = "user-"
+		}
+		autoStartKey, err := store.ReadState(startKey)
+		if err != nil && err != ipn.ErrStateNotExist {
+			return nil, fmt.Errorf("calling ReadState on state store: %w", err)
+		}
+		if len(autoStartKey) != 0 {
+			key := string(autoStartKey)
+			if strings.HasPrefix(key, stateKeyPrefix) {
+				stateKey = ipn.StateKey(key)
+			}
+		}
+	}
+
+	var knownProfiles []string
+	prfB, err := store.ReadState(ipn.KnownProfilesStateKey)
+	if err != nil && err != ipn.ErrStateNotExist {
+		return nil, fmt.Errorf("calling ReadState on state store: %w", err)
+	} else if err == nil {
+		if err := json.Unmarshal(prfB, &knownProfiles); err != nil {
+			return nil, fmt.Errorf("unmarshaling known profiles: %w", err)
+		}
+	}
+
+	if stateKey == "" && len(knownProfiles) == 0 && goos != "windows" {
+		knownProfiles = []string{"default"}
+		stateKey = "profile-default"
+
+		// No known profiles, see if there is a "_daemon" profile.
+		// If so, migrate it to the new format.
+		b, err := store.ReadState(ipn.GlobalDaemonStateKey)
```

Comment:
FWIW the Mac and iOS apps used a different state key for some (presumably historical/migration) reasons, so wed' need to read that too: https://github.com/tailscale/corp/blob/d96b92a627aeba0bec96591fc3e87ea7f8e66bb4/xcode/ipn-go-bridge/ipn.go#L245-L258

Reply:
Done